### PR TITLE
[IMP] Store binary fields into attachments

### PR DESCRIPTION
**Pull request by rco-odoo**
When enabled, store the content of binary fields into attachment.
For instance, `record.image` will be stored into an attachment `att` such that:

```
att.res_model = record._name
att.res_field = 'image'             # this is a new field
att.res_id = record.id
```

### CODE REVIEW

**Comment by rco-odoo** [comment]
@KangOl what do you think about this?

File: openerp/osv/fields.py [hunk 1]
```diff
@@ -583,35 +584,68 @@ class binary(_column):
...
+        if self.attachment:
+            # values are stored in attachments, retrieve them
+            atts = obj.pool['ir.attachment'].browse(cr, user, [], context)
+            domain = [
+                ('res_model', '=', obj._name),
+                ('res_field', '=', name),
+                ('res_id', 'in', ids),
+            ]
+            for att in atts.search(domain):
+                result[att.res_id] = att.datas
```
Comment (KangOl) [inline openerp/osv/fields.py]: Should handle `bin_size` flag

**Comment by KangOl** [comment]
I'm not fan of the mix old-api/new-api for fields.
It should still works with binary function fields of old-api

**Comment by rco-odoo** [comment]
@KangOl I am concerned by the mix old/new API, too.
The fact is that I cannot easily make it work for old-API binary function fields: they do not handle their storage by themselves, instead this is done in `BaseModel._store_set_values`...

File: openerp/osv/fields.py [hunk 2]
```diff
@@ -584,35 +585,69 @@ class binary(_column):
...
+            atts = obj.pool['ir.attachment'].browse(cr, user, [], context)
+            domain = [
+                ('res_model', '=', obj._name),
+                ('res_field', '=', name),
+                ('res_id', 'in', ids),
+            ]
+            for att in atts.search(domain):
+                # the 'bin_size' flag is handled by the field 'datas' itself
+                result[att.res_id] = att.datas
+        else:
```
Comment (KangOl) [inline openerp/osv/fields.py]: You should also postprocess attachment data to handle `bin_size`

**Comment by KangOl** [comment]
Also, nothing forbid to have more than one attachment for a field for a record.
You should had a unique constraint on (res_model, res_field, res_id)

**Comment by rco-odoo** [comment]
Effect on database: consider a new database with demo data and module `sale` installed.
 - before: db dump is 14.2Mb
 - after: db dump is 4.5Mb

File: openerp/osv/fields.py (hunk 2 above)
Comment (rco-odoo) [inline openerp/osv/fields.py]: `att` has the `bin_size` flag in its own context, therefore `att.datas` already handles it

**Comment by wtaferner** [comment]
:+1:

**Comment by rco-odoo** [comment]
@KangOl by construction, the unique constraint will be satisfied.
If it is not (you screwed up stuff in your db), the code will not crash but consider one of the attachments.
Enforcing the constraint will:
 - potentially make stuff crash without the user knowing what to do;
 - cost performance.

**Comment by rco-odoo** [comment]
Here are some extra data for a database created with `odoo.py -d DB -i sale`:

Case | DB Dump (gzipped) | Filestore
---|---|---
Binary fields stored in database | 14M (4.7M) | 20K
Binary fields stored in attachments | 4.4M (571K) | 5.3M

The database backup is considerably smaller (almost 90% smaller when gzipped).
The filestore is larger, but its backups are more incremental (rsnapshot).

**Comment by KangOl** [comment]
Note: runbot is red.

**Comment by KangOl** [comment]
By default, TransientModels should not store binary fields in attachments.

**Comment by rco-odoo** [comment]
@KangOl I fixed the case of TransientModel and others by not making `attachment=True` the default.
The binary fields stored in attachments must be base64-encoded, and some binary fields are not.

**Comment by rco-odoo** [comment]
Branch is now :white_check_mark: on runbot!

**Comment by KangOl** [comment]
Seems correct.

**Pull request merged**